File: src/engines/cad/builder.py
"""
Geometry generation step of the CAD workflow (PRD §14):

    Engineering Parameters -> CAD IR -> Constraint Generation ->
    Geometry Generation -> Geometry Validation -> ...

This module is the "Geometry Generation" step for the built-in
quadcopter_frame primitive.
"""

from __future__ import annotations

import math

from src.engines.cad.ir import QuadcopterFrameIR
from src.engines.cad.primitives import Mesh, box

ARM_ANGLES_DEG = (45.0, 135.0, 225.0, 315.0)  # X configuration


def build_quadcopter_frame(ir: QuadcopterFrameIR) -> Mesh:
    p = ir.parameters
    plate_size = p["center_plate_size"]
    thickness = p["plate_thickness"]
    arm_width = p["arm_width"]
    overall = p["overall_size"]
    mount_d = p["motor_mount_diameter"]

    mesh = box(center=(0, 0, 0), size=(plate_size, plate_size, thickness))

    start_r = (plate_size / 2) * math.sqrt(2)
    end_r = overall / 2
    arm_length = end_r - start_r

    for angle in ARM_ANGLES_DEG:
        theta = math.radians(angle)
        center_r = (start_r + end_r) / 2
        arm_center = (center_r * math.cos(theta), center_r * math.sin(theta), 0.0)
        mesh.extend(
            box(
                center=arm_center,
                size=(arm_length, arm_width, thickness),
                rotation_z_deg=angle,
            )
        )

        motor_center = (end_r * math.cos(theta), end_r * math.sin(theta), 0.0)
        boss_side = mount_d * 1.4
        mesh.extend(
            box(
                center=motor_center,
                size=(boss_side, boss_side, thickness * 1.5),
                rotation_z_deg=angle,
            )
        )

    return mesh

File: app/native/src/cad/Builder.cpp
#include "trinity/cad/Builder.hpp"

#include <cmath>

namespace trinity::cad {

const double kArmAnglesDeg[4] = {45.0, 135.0, 225.0, 315.0};

Mesh buildQuadcopterFrame(const FrameParams& params) {
    const auto& p = params.parameters;
    const double plateSize = p.at("center_plate_size");
    const double thickness = p.at("plate_thickness");
    const double armWidth = p.at("arm_width");
    const double overall = p.at("overall_size");
    const double mountD = p.at("motor_mount_diameter");

    constexpr double kPi = 3.14159265358979323846;
    Mesh mesh = box({0.0, 0.0, 0.0}, {plateSize, plateSize, thickness});

    const double startR = (plateSize / 2.0) * std::sqrt(2.0);
    const double endR = overall / 2.0;
    const double armLength = endR - startR;

    for (double angle : kArmAnglesDeg) {
        const double theta = angle * kPi / 180.0;
        const double centerR = (startR + endR) / 2.0;
        mesh.extend(box({centerR * std::cos(theta), centerR * std::sin(theta), 0.0},
                        {armLength, armWidth, thickness}, angle));

        const double bossSide = mountD * 1.4;
        mesh.extend(box({endR * std::cos(theta), endR * std::sin(theta), 0.0},
                        {bossSide, bossSide, thickness * 1.5}, angle));
    }
    return mesh;
}

}  // namespace trinity::cad
